### PR TITLE
WIP: initial sso updates

### PR DESCRIPTION
**Pull request by rhoads-zach**

### CODE REVIEW

**Comment by rhoads-zach** [comment]
initial fixes for #718 #720 #722 and #730

**Comment by rhoads-zach** [comment]
FYI @heiko-braun mind taking a look?

File: docs/topics/secured-mission-common-interaction.adoc [hunk 1]
```diff
@@ -2,33 +2,103 @@
 
 The Secured booster provides a default HTTP endpoint that accepts `GET` requests if the caller is authenticated and authorized. The `sso` directory of the booster contains a Java client you can use to interact with the Secured booster. The client first authenticates against the {RHSSO} server and then performs a `GET` request against the Secured booster using the access token returned by the authentication step.
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: The `sso` directory doesn't exist anymore.

File: docs/topics/secured-mission-common-interaction.adoc [hunk 2]
```diff
@@ -2,33 +2,103 @@
...
 
-The Java client is packaged as an executable JAR. It is located at `target/sso-client.jar` and takes arguments from the CLI. You can use the `--help` argument to see a full list of available arguments.
+[[secured-endpoint]]
+== The Secured Endpoint
+When using a client to interact with the booster, you must specify the Secured booster endpoint, which is the Greeting servic). You can get a list of the endpoints using the `oc get routes` command. The following table shows sample output from the `oc get routes` where you can obtain the Secured booster endpoint name (`<SERVICE_HOST>`).
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: Probably reword to mention that this way, you figure out _both_ the URL of the SSO server and of the secured application.

File: docs/topics/secured-mission-common-interaction.adoc [hunk 3]
```diff
@@ -45,3 +115,16 @@ NOTE: This is the master realm public key from the Secured booster's {RHSSO} ser
...
+
+== Using the Web Interface
+The secured endpoint also contains a web interface to interact with it. Access it using the xref:secured-endpoint[endpoint address].
+
+When using the web interface, you will go through the following steps to see how security is enforced, how authentication happens, and how the token is requested and attached to an authenticated invocation.
+
+. Open the web interface and click on _Invoke_. This sends an unauthenticated request to the Greeting service. 
+. The services responds with an `HTTP 401 Unauthenticated` status code.
+. Click _Login_ to authenticate yourself against {RHSSO}. This redirects you to the SSO server and requests your credentials. 
+. Log in as _Alice_.
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: Do we mention the username/password combination somewhere?

File: docs/topics/secured-mission-common-interaction.adoc [hunk 4]
```diff
@@ -45,3 +115,16 @@ NOTE: This is the master realm public key from the Secured booster's {RHSSO} ser
...
+. Once authenticated you are redirected to the web interface, which includes a bearer token shown in the command line output at the bottom of the page.
+. Click _Invoke_ again to access the Greeting service.
+. Confirm there is no exception and JSON response payload displayed. This means the service accepted your bearer token and you are authorized access to the Greeting service.
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: This should continue with:

- logout
- try invoking again
- login as admin/admin
- try invoking again -- will see different error because the admin user, while authenticated, isn't authorized

File: docs/topics/secured-mission-intro.adoc [hunk 5]
```diff
@@ -1,15 +1,65 @@
...
+
+
+a|
+The configured **clients accept any callback url**.
+
+a|
+To avoid having a custom configuration per runtime we avoid the verification of the callback as mandated by the OAuth2 specification.
+
+a|
+A application specific callback URL should be provided with a valid domain name.
```
Comment (Ladicek) [inline docs/topics/secured-mission-intro.adoc]: __An__ application specific ..., right?

File: docs/topics/secured-mission-intro.adoc [hunk 6]
```diff
@@ -1,15 +1,65 @@
...
+
+a|
+**Clients do not require SSL/TLS**.
+
+a|
+The boosters are simplified by not requiring certificates generated for each runtime.
+
+a|
+In production a secure application should use HTTPS over plain HTTP.
```
Comment (Ladicek) [inline docs/topics/secured-mission-intro.adoc]: Instead of "over", could we use "rather than"? The "over" word is often used for protocol layering, which isn't the case here.

File: docs/topics/secured-mission-intro.adoc [hunk 7]
```diff
@@ -1,15 +1,65 @@
...
+The configured **clients accept any callback url**.
+
+a|
+To avoid having a custom configuration per runtime we avoid the verification of the callback as mandated by the OAuth2 specification.
+
+a|
+A application specific callback URL should be provided with a valid domain name.
+
+a|
+**Clients do not require SSL/TLS**.
```
Comment (Ladicek) [inline docs/topics/secured-mission-intro.adoc]: I'd maybe reword or add that "the secured applications aren't exposed over HTTPS".

File: docs/topics/secured-mission-intro.adoc [hunk 8]
```diff
@@ -1,15 +1,65 @@
...
+In production a secure application should use HTTPS over plain HTTP.
+
+a|
+**The token timeout has been increased to 10 minutes from the default of 1 minute.**
+
+a|
+Provides a better user experience when working with the command line examples
+
+a|
+From a security perspective, the window an attacker would have to guess the access token is extended. It is recommended to keep this window short as the rotation creates entropy to guess the current token.
```
Comment (Ladicek) [inline docs/topics/secured-mission-intro.adoc]: "rotation creates entropy to guess the current token" is a mishmash of words :-) I'd reword to "it is recommended to keep this window short, which makes it much harder for a potential attacker to guess the current token".

File: docs/topics/secured-mission-vertx-adapter.adoc [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+JsonObject config = new JsonObject()
+  	.put("realm", System.getenv("REALM")) <1>
+  	.put("public-key", System.getenv("REALM_PUBLIC_KEY")) <3>
+  	.put("auth-server-url", System.getenv("SSO_AUTH_SERVER_URL")) <4>
+  	.put("ssl-required", "external")
+  	.put("resource", System.getenv("CLIENT_ID")) <2>
+  	.put("credentials", new JsonObject()
+    	.put("secret", System.getenv("SECRET")))
+  	// since we're consuming keycloak JWTs we need to locate the permission claims in the token
+  	.put("permissionsClaimKey", "realm_access/roles"); <5>
```
Comment (Ladicek) [inline docs/topics/secured-mission-vertx-adapter.adoc]: Why 1, 3, 4, 2, 5?

Comment (tradej) [reply docs/topics/secured-mission-vertx-adapter.adoc]: In the source, they are just identifiers. AsciiDoc replaces the numbers so that it's 1, 2, 3, 4, 5.

Comment (Ladicek) [reply docs/topics/secured-mission-vertx-adapter.adoc]: Ah interesting :-)

**Comment by rhoads-zach** [comment]
thank you for the feedback @Ladicek, fixed,

File: docs/topics/secured-mission-common-interaction.adoc [hunk 10]
```diff
@@ -45,3 +115,20 @@ NOTE: This is the master realm public key from the Secured booster's {RHSSO} ser
...
+. The services responds with an `HTTP 401 Unauthenticated` status code.
+. Click _Login_ to authenticate yourself against {RHSSO}. This redirects you to the SSO server and requests your credentials. 
+. Log in as xref:sso-users[the _Alice_ user].
+. Once authenticated you are redirected to the web interface, which includes a bearer token shown in the command line output at the bottom of the page.
+. Click _Invoke_ again to access the Greeting service.
+. Confirm there is no exception and JSON response payload displayed. This means the service accepted your bearer token and you are authorized access to the Greeting service.
+. Logout.
+. Click on _Invoke_. Confirm this sends an unauthenticated request to the Greeting service. 
+. Click _Login_ and log in as xref:sso-users[the _admin_ user].
+. Click on _Invoke_. Confirm this sends an authenticated request to the Greeting service, but you receive an error relating to the user being unauthorized.
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: Should probably mention, similarly to how we mention `HTTP 401` above, that this is `HTTP 403 Forbidden`.

**Comment by Ladicek** [comment]
Thanks @rhoads-zach, added just one more minor comment, otherwise LGTM. I'd welcome review from @heiko-braun / @cescoffier / @pmlopes, too.

**Comment by rhoads-zach** [comment]
thanks @Ladicek fixed and merging.

File: docs/topics/secured-mission-common-interaction.adoc [hunk 11]
```diff
@@ -45,3 +115,21 @@ NOTE: This is the master realm public key from the Secured booster's {RHSSO} ser
...
+. Click _Login_ to authenticate yourself against {RHSSO}. This redirects you to the SSO server and requests your credentials. 
+. Log in as xref:sso-users[the _Alice_ user].
+. Once authenticated you are redirected to the web interface, which includes a bearer token shown in the command line output at the bottom of the page.
+. Click _Invoke_ again to access the Greeting service.
+. Confirm there is no exception and JSON response payload displayed. This means the service accepted your bearer token and you are authorized access to the Greeting service.
+. Logout.
+. Click on _Invoke_. Confirm this sends an unauthenticated request to the Greeting service. 
+. Click _Login_ and log in as xref:sso-users[the _admin_ user].
+. Click on _Invoke_. 
+. Confirm the service responds with an `HTTP 403 Unauthorized` status code since the _admin_ user is not authorized to access the Greeting service.
```
Comment (Ladicek) [inline docs/topics/secured-mission-common-interaction.adoc]: One more issue here. previously, the 1st item in thist list had `401 Unauthorized`, and I asked you to add `403 Forbidden` to this last item of the list. What you did was changing `401 Unauthorized` to `401 Unauthenticated` and added `403 Unauthorized`.

It's true that 401 meands unauthenticated and 403 means unauthorized, but the typical status messages are `401 Unauthorized` and `403 Forbidden`. Could you please change that back? Thanks.

Comment (rhoads-zach) [reply docs/topics/secured-mission-common-interaction.adoc]: sorry, missed this when merging. i'll file a follow up PR @Ladicek

Comment (Ladicek) [reply docs/topics/secured-mission-common-interaction.adoc]: Sure, np.

**Comment by rhoads-zach** [comment]
@Ladicek fixed in #733
